photon-test/src/main/scala/com/linkedin/photon/ml/test/TestTemplateWithTmpDir.scala

```diff
@@ -24,15 +27,25 @@ import org.testng.annotations.AfterMethod
  * @author dpeng
  */
 trait TestTemplateWithTmpDir {
+
+  private val tmpDirs = new mutable.ArrayBuffer[String]()
+
   /**
    * Return the temporary directory as a string.
    * @return the temporary directory as a string.
    */
-  def getTmpDir: String = TestTemplateWithTmpDir._tmpDirThreadLocal.get()
+  def getTmpDir: String = {
+    val tmpDirName = TestTemplateWithTmpDir._tmpDirThreadLocal.get()
+    tmpDirs.synchronized {
+      tmpDirs += tmpDirName
+    }
+    tmpDirName
+  }
 
   @AfterMethod
```

Comment:
I think in multithreading scenario, AfterMethod should probably changed to AfterClass. Currently, we are setting parallel level of "methods" (https://github.com/linkedin/photon-ml/blob/master/build.gradle#L115). Different methods in the same class will share the same class variables. These directory should only be deleted once. For "methods" parallelism, AfterClass is only executed once, while AfterMethod will be executed multiple times (by different threads). So I think this should be AfterClass.